Patch by Martin Pohlack <mp26@os.inf.tu-dresden.de> fixing a long
standing bug where Shift+PgUp and Shift+PgDn key strokes (possibly
among others) do not work properly when mouse cursor isn’t over an
inside of a tab.

The bug affects tabbedex as well as the original tabbed extension.  To
reproduce:
1. Start urxvt with tabbed or tabbedex extension.
2. Run ‘seq $(($LINES * 5))’ or similar command to produce multiple
   screens of text.
3. Focus the terminal window and move mouse pointer over it.
4. Press Shift+PgUp and observe content scroll up.
5. Move mouse pointer outside of the terminal window while not
   keeping focus on the window.
6. Press Shift+PgUp or Shift+PgDn.
Expected: Content scrolls up or down (as in step 4).
Observed: Nothing happens.

If you are affected by this bug, this patch may help.

The patch has been originally developed for rxvt-unicode 9.19 but
applies (with a small offsets) at rxvt-unicode 2.21 as well.

To apply, download latest rxvt-unicode release and before building
from source run:

  patch -p1 </path/to/this/file

diff -u rxvt-unicode-9.19.orig/src/command.C rxvt-unicode-9.19/src/command.C
--- rxvt-unicode-9.19.orig/src/command.C	2013-05-06 07:37:06.000000000 +0200
+++ rxvt-unicode-9.19/src/command.C	2014-03-02 20:28:17.045766536 +0100
@@ -410,6 +410,10 @@
   int valid_keysym;
   char kbuf[KBUFSZ];
 
+  // allow early handling of events for plugins
+  if (HOOK_INVOKE ((this, HOOK_KEY_PRESS, DT_XEVENT, &ev, DT_END)))
+    return;
+
 #if ISO_14755
   if (iso14755buf & ISO_14755_52)
     return;
@@ -864,6 +902,10 @@
 void ecb_cold
 rxvt_term::key_release (XKeyEvent &ev)
 {
+  // allow early handling of events for plugins
+  if (HOOK_INVOKE ((this, HOOK_KEY_RELEASE, DT_XEVENT, &ev, DT_END)))
+    return;
+
 #if (MOUSE_WHEEL && MOUSE_SLIP_WHEELING) || ISO_14755 || ENABLE_PERL
   KeySym keysym;
 
diff -u rxvt-unicode-9.19.orig/src/rxvtperl.xs rxvt-unicode-9.19/src/rxvtperl.xs
--- rxvt-unicode-9.19.orig/src/rxvtperl.xs	2013-10-03 02:55:37.000000000 +0200
+++ rxvt-unicode-9.19/src/rxvtperl.xs	2014-03-02 20:25:19.164884465 +0100
@@ -1535,6 +1535,9 @@
           THIS->key_release (xkey);
         else
           THIS->key_press (xkey);
+
+        // similar refresh check as in standalone x_cb() function
+        THIS->refresh_check();
 }
 
 void
